.github/workflows/build-pipeline.yml

```diff
@@ -65,4 +62,35 @@ jobs:
             "text": ":warning: GitHub Action cancelled by ${{ github.actor }}"
           }
       env:
-        SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
\ No newline at end of file
+        SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
+
+  docker:
+    name: Push Docker image to Docker Hub
+    runs-on: ubuntu-latest      
+    needs: [build]
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v4
+
+      - name: Log in to Docker Hub
+        uses: docker/login-action@f4ef78c080cd8ba55a85445d5b36e214a81df20a
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+
+      - name: Build and push Docker image
+        id: push
+        uses: docker/build-push-action@3b5e8027fcad23fda98b2e3ac259d8d67585f671
+        with:
+          context: .
+          file: ./Dockerfile
+          push: true
+          target: production
+          tags: ${{ secrets.DOCKER_USERNAME }}/todo-app:production
+          cache-from: |
+            ${{ secrets.DOCKER_USERNAME }}/todo-app:production
+            type=registry,ref=${{ secrets.DOCKER_USERNAME }}/todo-app:production
```

Comment:
Given that my local one builds in about 3s don't think this has done much. Any ideas why? From the logs it looks like it's getting something
| Before | After |
|---------|-------|
|![image](https://github.com/user-attachments/assets/6583967f-18a0-47d6-a6fc-983f82c19e5b)|![image](https://github.com/user-attachments/assets/c0332992-8f00-4778-9450-fe5b3ca7e138)|

![image](https://github.com/user-attachments/assets/6f14e303-493f-4d11-b7ce-5cfb980ccc7c)

Reply:
Hmm, doesn't look like it's achieving much I agree!

It _could_ be because we're using a multi-stage build, and it's defaulting to ["cache mode: min"](https://docs.docker.com/build/cache/backends/#cache-mode):
> In min cache mode (the default), only layers that are exported into the resulting image are cached, while in max cache mode, all layers are cached, even those of intermediate steps.

It's also worth noting that this build-push action suggests that you [setup buildx](https://github.com/docker/buildx) first to ensure they can use all the features.

The solution I've used before here is using the inbuilt GitHub actions cache type, which also relies on [buildx](https://github.com/docker/buildx), something like:
```yml
        - uses: docker/setup-buildx-action@v3
        - uses: docker/build-push-action@v5
          with:
            context: .
            ...
            cache-from: type=gha
            cache-to: type=gha,mode=max
```

(and if you're interested, [here's an example repo I used to explore](https://github.com/JackMead/docker-caching/blob/main/.github/workflows/caching-test.yml) - the suggested approach is primarily the third of the three jobs)

Reply:
Yep that's much better, thanks!
![image](https://github.com/user-attachments/assets/d38463e5-e3af-49da-b083-785bcc0bd48b)